nighthawk/test/benchmark_http_client_test.cc

```diff
@@ -0,0 +1,410 @@
+#include "common/network/dns_impl.h"
+#include "common/network/utility.h"
+#include "common/runtime/runtime_impl.h"
+#include "common/stats/isolated_store_impl.h"
+#include "common/thread_local/thread_local_impl.h"
+
+#include "nighthawk/source/client/benchmark_client_impl.h"
+#include "nighthawk/source/common/platform_util_impl.h"
+#include "nighthawk/source/common/rate_limiter_impl.h"
+#include "nighthawk/source/common/sequencer_impl.h"
+#include "nighthawk/source/common/statistic_impl.h"
+
+#include "envoy/thread_local/thread_local.h"
+#include "test/mocks/runtime/mocks.h"
+#include "test/mocks/thread_local/mocks.h"
+
+#include "test/integration/http_integration.h"
+
+#include "test/integration/integration.h"
+#include "test/integration/utility.h"
+#include "test/server/utility.h"
+#include "test/test_common/utility.h"
+
+using namespace std::chrono_literals;
+
+namespace Nighthawk {
+
+class BenchmarkClientTestBase : public Envoy::BaseIntegrationTest,
+                                public testing::TestWithParam<Envoy::Network::Address::IpVersion> {
+public:
+  BenchmarkClientTestBase()
+      : Envoy::BaseIntegrationTest(GetParam(), realTime(), BenchmarkClientTestBase::envoy_config),
+        api_(thread_factory_, store_, timeSystem(), file_system_),
+        dispatcher_(api_.allocateDispatcher()) {}
+
+  static void SetUpTestCase() {
+    Envoy::Filesystem::InstanceImplPosix file_system;
+    envoy_config = file_system.fileReadToEnd(Envoy::TestEnvironment::runfilesPath(
+        "nighthawk/test/test_data/benchmark_http_client_test_envoy.yaml"));
+    envoy_config = Envoy::TestEnvironment::substitute(envoy_config);
+  }
+
+  void SetUp() override {
+    ares_library_init(ARES_LIB_INIT_ALL);
+    Envoy::Event::Libevent::Global::initialize();
+  }
+
+  void TearDown() override {
+    if (client_.get() != nullptr) {
+      client_->terminate();
+    }
+    test_server_.reset();
+    fake_upstreams_.clear();
+    tls_.shutdownGlobalThreading();
+    ares_library_cleanup();
+  }
+
+  uint32_t getTestServerHostAndPort() { return lookupPort("listener_0"); }
+
+  void testBasicFunctionality(const std::string uriPath, uint64_t max_pending,
+                              uint64_t connection_limit, bool use_h2, uint64_t amount_of_request) {
+    setupBenchmarkClient(uriPath, use_h2);
+
+    client_->setConnectionTimeout(10s);
+    client_->setMaxPendingRequests(max_pending);
+    client_->setConnectionLimit(connection_limit);
+    EXPECT_TRUE(client_->initialize(runtime_));
+
+    uint64_t amount = amount_of_request;
+    uint64_t inflight_response_count = 0;
+
+    std::function<void()> f = [this, &inflight_response_count]() {
+      --inflight_response_count;
+      if (inflight_response_count == 0) {
+        dispatcher_->exit();
+      }
+    };
+
+    for (uint64_t i = 0; i < amount; i++) {
+      if (client_->tryStartOne(f)) {
+        inflight_response_count++;
+      }
+    }
+
+    EXPECT_EQ(max_pending, inflight_response_count);
+
```

Comment:
Nit: prefer not to have this much superfluous whitepace. Google style prefers compactness.